Message loss testcase

jhooda opened the pull request:
@ennru @seglo I have modified the "message loss" test case to reflect the following deployment conditions:

- Each consumer can be a separate process

- The consumers can leave or join the group abruptly (simulated using akka.stream.SharedKillSwitch)

- Expectation is that there can be message duplication but there shouldn't be any message loss

I observed message loss approximately once in every 30 test runs.

Below are few observations that I noted using extra debug logs in both kafka-client library and alpaka-kafka library:

- The lost messages are always in a multiple of the following configuration:
`   ConsumerConfig.MAX_POLL_RECORDS_CONFIG, "20"
`

- The lost messages were last seen at the following location 
```
File: core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala:612
            ...
            stageActorRef ! Messages(req.requestId, messages)
            ...
``` 

- But they do not show up at this location
```
File:core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala:
...
  protected def messageHandling: PartialFunction[(ActorRef, Any), Unit] = {     
    case (_, msg: KafkaConsumerActor.Internal.Messages[K, V]) =>                
      requested = false                                                         
      buffer = buffer ++ msg.messages                                           
      pump()                                                                    
    case (_, Status.Failure(e)) =>                                              
      failStage(e)                                                              
    case (_, Terminated(ref)) if ref == consumerActor =>                        
      failStage(new ConsumerFailed)                                             
  }
...
```

- The lost message object is not directly present in DeadLetterListener. But there are plenty of other messages in DeadLetterListener which may be hiding the lost messages

- The messages are lost during Rebalance event (e.g., SharedKillSwitch trigger)

Few other observations that r not directly relevant but are interesting.

- Although the abrupt failure is initiated by SharedKillSwitch the consumers do not abandon the currently processing message batch (20 messages per batch). Why? Are we suppressing the kill in some fashion? Is there a way to not suppress it?

- There is excessive draining seen in kafka-client trace logs (issue #549). Almost all message batches are fetched twice in which one batch-fetch is discarded/drained. This may explain the excessive network bandwidth seen in the mentioned issue.

```
Project: kafka_2.12_2.4.0
File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java
...
log.trace("Draining fetched records for partition {}", completedFetch.partition)
...
```

ennru committed:
Reduce logging

ennru committed:
Reduce logging more

ennru committed:
Log duplicated messages at the end

ennru committed:
Wait for rebalance

seglo committed:
Drain sub sources. Cancel when all sub sources msgs committed.

seglo committed:
Remove unused producer defaults override

jhooda committed:
Added a test case to validate no message loss during rebalance

jhooda committed:
Merge branch 'message-loss-testcase' of github.com:jhooda/alpakka-kafka into message-loss-testcase

jhooda committed:
Modified message loss test case to reflect deployment conditions

in which each one the three consumers can run as independent process,
and join or leave the consumer group abruptly. Expectations are messages
shouldn't be lost under such rebalance conditions

seglo commented on tests/src/test/scala/akka/kafka/scaladsl/RebalanceSpec.scala:
```diff
@@ -256,100 +264,107 @@ class RebalanceSpec extends SpecBase with TestcontainersKafkaLike with Inside {
-        stream
-        // end the sub source once all messages have been committed.  this actually isn't that helpful because the
-        // partitioned source will just re-emit the sub source if it shuts down anyway
-          .takeWhile { batch =>
-            val lastOffset = batch.offsets.head._2
-            log.debug(s"Batch committed offset: $lastOffset for tp ${batch.offsets.head._1.topicPartition}")
-            if (lastOffset == 999) {
-              val count = partitionProcessedCount.incrementAndGet()
-              log.debug(
-                s"Processing complete for ${batch.offsets.head._1.topicPartition}. Incrementing partition processed count ($count)."
-              )
-              // if all sub sources have been fully processed then cancel the outer stream
-              if (count == (topicCount * partitionCount)) {
-                log.debug("All partitions have been processed")
-                outerStream.map(_.cancel())
-              }
-            }
-            lastOffset < 999
-          }
```
This logic is important to assert make sure a partition of messages is fully processed. If you don't do this (or something like it) then there's no guarantee that any consumer will have time to process all messages before you complete them at the end of the test.  Instead of relying on `Thread.sleeps` you should add blocking (with a timeout) to make sure all partitions are fully processed.  This check can be done by observing that the last offset is encountered.

jhooda replied:
@seglo I have kind of simplified the assertion using a promise map (322-325,Green) and then waiting on the future at line 470.

The missing messages are never from the end of the partition message batch, but are during the forced rebalance trigger points (e.g., line 432 and 440). The failure frequency is also quite low (approximately 1 failure in 30 runs), but it is there:-) I was simplifying the test so that it can become part of the RebalanceSpec by using a lower value for
`val perPartitionMessageCount = 10 \\ currently 1000`,
but then I noticed the failure. I added a lot of debug statements and came to conclusion that the lost batch is sent at

```
File: core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala:612
            ...
            stageActorRef ! Messages(req.requestId, messages)
            ...

```
But is never received at 

```
File:core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala:
...
  protected def messageHandling: PartialFunction[(ActorRef, Any), Unit] = {     
    case (_, msg: KafkaConsumerActor.Internal.Messages[K, V]) =>                
      requested = false                                                         
      buffer = buffer ++ msg.messages                                           
      pump()                                                                    
    case (_, Status.Failure(e)) =>                                              
      failStage(e)                                                              
    case (_, Terminated(ref)) if ref == consumerActor =>                        
      failStage(new ConsumerFailed)                                             
  }
...

```
After comparing the message blocks before and after the lost message block I found that the messages were sent to a defunct/dead stageActorRef (the referred code block above), which is why they never made it, but the offset logic in Fetcher.java was already advanced to the next block.

seglo reviewed (commented):
Hi. I think this test is flawed without asserting that the last offset per partition is encountered (see inline comment). This was the main change I made to your original tests.

It's worth noting that we have no received any other reports from Lightbend clients or opensource users of missing offsets when using the at-least-once semantics. This test is quite large and time-consuming to troubleshoot. If you find an easier way to recreate the issue you're encountering then we'll certainly work with you to address it, but as it stands it's difficult to justify the time to continue to review it thoroughly.

I admit Alpakka Kafka and Akka Streams makes it easy to "shoot yourself in the foot" and create conditions where message loss can occur. If you're a Lightbend client please feel free to escalate this conversation through our support portal and maybe we can provide direct assistance and review the actual application you're observing these issues with.

jhooda committed:
added a fix for message-batch loss during re-balancing

Issue: When there are multiple [KafkaConsumerActor]s they may get
assigned an orthoginal set of topic-partitions, backed by independent
[StageActor]s during group re-balancing. All [StageActor]s receive
a feed of message streams from [KafkaConsumerActor]. At the other end,
[KafkaConsumerActor] receives a primary message stream from a shared
message [Fetcher] as shown below:

[Kafka Broker]-->[Fetcher]--+>[KafkaConsumerActor]--+>[StageActor]

During normal message flow [Fetcher] saves the next message offset
internally and uses that as a reference offset to deliver the next
batch of messages to [KafkaConsumerActor]. However, during
re-balancing the in-progress message batch can be lost as shown below

[Fetcher]--+>[KafkaConsumerActor]--+>[*defuct* StageActor]

Since the entire message stream is asynchornous, [Fetcher] doesn't
always know about the lost message batch, and instead delivers the
next message batch to post-rebalance [new StageActor]

[Fetcher]-->[KafkaConsumerActor]-->[New StageActor]

This fix resets the seek position to the last commit offset, whenever
a [new StageActor] is initialized.

How to reproduce the issue?
In RebalanceSpec change perPartitionMessageCount value from 10 to 1000
and run about 50 itertations.

jhooda commented:
@seglo I have submitted a patch to fix the mentioned issue. Will it be possible to merge the change in main branch? I have ran many round of tests after the fix and it essentially been addressed.

seglo commented:
@jhooda Thanks for keeping at it.  Your explanation in https://github.com/ennru/alpakka-kafka/pull/3/commits/1dda36f3c1554543d4f47e7e9e79170f452e88bd makes sense to me. If I understand correctly, you're asserting this message loss only happens with partioned `Sources` and when a rebalance re-assigns a partition back to the same consumer group member?  The previous partition `Source` is cancelled, but that may occur before inflight messages are processed and committed.  Meanwhile, since the Kafka Consumer group member is the same consumer instance it will continue returning messages from the last polled offset for the partition.

My understanding was that during a rebalance the Consumer would internally reset its position to the last committed offset, even if it previously had that partition.  I've seen this behaviour when using Kafka transactions.

I think we should be able to test this scenario more reliably and succinctly by manually assigning partitions (to assure they're assigned back to the same consumer group members), and introducing a blocking operation in the partition `Source` so that it can't commit offsets while a rebalance is occurring.  Does that make sense?  I don't think there's much value in merging a test that only resolves the issue occassionally.

To proceed I would recommend squashing all this work into a single commit (make sure to preserve your explanation from https://github.com/ennru/alpakka-kafka/pull/3/commits/1dda36f3c1554543d4f47e7e9e79170f452e88bd), rebase on master, create a test that reliably reproduces the issue, and create a new PR to https://github.com/akka/alpakka-kafka.  If you want, I can help you with the test, but I'll need some time.

jhooda commented:
@seglo Yes your assessment is point on. Thanks for condensing it down. The consumer is essentially not resetting the partition offset when it is resuming a partition post rebalance. Now that we know the issue, a test along your suggestion will be able to capture it nicely. I can work on it and submit a PR.

seglo commented:
Great.  Thanks @jhooda!

jhooda commented:
Hi @seglo @ennru 

Greetings! It's been a long time since we discussed this issue. (I had to take a break for personal reasons.) I've submitted a new pull request as per our above discussion. My earlier suggested change didn't quite work and I had to come out with a newer fix. Please review https://github.com/akka/alpakka-kafka/pull/1263 and let me know if it make sense. In my dev env, I'm able to re-produce the issue in about 85/100 test runs (since the entire system is messages based it is quite tricky to get the messaging order right that exposes this issue).

One can run the test as

`sbt 'tests/testOnly akka.kafka.scaladsl.RebalanceExtSpec -- -z "no messages should be lost when two consumers consume from one topic and two partitions and one consumer aborts mid-stream"'`

Thanks,
Jagbir
